distribution_metadata.c

```diff
@@ -419,28 +370,30 @@ PartitionType(Oid distributedTableId)
 bool
 IsDistributedTable(Oid tableId)
 {
+	Oid metadataNamespaceOid = get_namespace_oid("pgs_distribution_metadata", false);
+	Oid tableNamespaceOid = get_rel_namespace(tableId);
 	bool isDistributedTable = false;
-	RangeVar *heapRangeVar = NULL;
-	Relation heapRelation = NULL;
-	HeapScanDesc scanDesc = NULL;
-	const int scanKeyCount = 1;
-	ScanKeyData scanKey[scanKeyCount];
-	HeapTuple heapTuple = NULL;
-
-	heapRangeVar = makeRangeVar(METADATA_SCHEMA_NAME, PARTITION_TABLE_NAME, -1);
-	heapRelation = relation_openrv(heapRangeVar, AccessShareLock);
+	Oid argTypes[] = { OIDOID };
+	Datum argValues[] = { ObjectIdGetDatum(tableId) };
+	const int argCount = sizeof(argValues) / sizeof(argValues[0]);
+	int spiStatus = 0;
 
-	ScanKeyInit(&scanKey[0], ATTR_NUM_PARTITION_RELATION_ID, InvalidStrategy,
-				F_OIDEQ, ObjectIdGetDatum(tableId));
+	if (tableNamespaceOid == metadataNamespaceOid)
```

Comment:
> What is this check for? Does this mean if the table is in in the pgs_distribution_metadata namespace, it isn't considered distributed?

Exactly.

SPI exercises the full PostgreSQL machinery, including `pg_shard` itself. `pg_shard` starts every query by iterating over relations and determining whether they are distributed. If none are, we use the PostgreSQL planner/executor.

If we have a query like `SELECT * FROM foo`, we'll check whether `foo` is distributed. That requires making a `SELECT` query against `pgs_distribution_metadata.partition`. But then `pg_shard` kicks in and wants to find out whether `pgs_distribution_metadata` is distributed… an infinite loop.

By inserting a short-circuit base case for this recursion, we can escape the infinite loop. I see no reason to expect any of the distribution metadata tables will be distributed, so this is an easy way out of the recursion.